### ISSUE TITLE
Does not work when the canvas is not in Screen Space - Overlay

### ISSUE THREAD

**Issue by LodenDarkStar** [zloedi/cui_color_picker#1]
This issue is due to how you are testing for the mouse position.

Here is the modification from my variant that corrected the problem
`private static bool GetLocalMouse(RectTransform rt, out Vector2 result)
        {
            RectTransformUtility.ScreenPointToLocalPointInRectangle(rt, Input.mousePosition, Camera.main, out mousePosition);
            result.x = Mathf.Clamp(mousePosition.x, rt.rect.min.x, rt.rect.max.x);
            result.y = Mathf.Clamp(mousePosition.y, rt.rect.min.y, rt.rect.max.y);
            Debug.Log(rt.rect.Contains(mousePosition));
            return rt.rect.Contains(mousePosition);
        }`

Note I have also changed most of the code at this point to remove any local declrations to variables within funcitons that can be called on an update loop such as the GetLocalMouse function hince the new private static mousePosition that use as opposed to var mp as it was in your code branch.

**Comment by zloedi** [comment]
Thanks for your fix there. Unfortunately when i use ScreenPointToLocalPointInRectangle, the mouse input handling breaks for Canvases with Screen Space - Overlay Render Mode. Also the code breaks if no camera is tagged as MainCamera. We can either:
    Detect the Render Mode for the parent canvas and the attached Render Camera / Event Camera.
    Supply the camera to the picker script in some way.
    Let the user supply the mouse positition.
    More?

**Comment by LodenDarkStar** [comment]
Yes my usual preference is to on start of the GO see if a camera was provided by the dev e.g. in a Camera field if not I set the field to main camera if that is null I go looking for the top most event camera.

As to handling events you could get around that by leveraging the event data already coming from the UI controls which represent the various inputs e.g. the UnityEngine.UI.Textures for the saturation and hue where in you would handle drag event on each repsectivly and where that was meaningful you would pass that change up to the parent color picker to process. The advantage here is that you spliting that logic out to each respective element thus its a bit easier to modify for the dev without impacting the picker its self the draw back is its yet anotehr behaviour script to maintain but considering at current its all in one thats not to bad.

**Comment by LodenDarkStar** [comment]
Ah forgot to comment on use when in screen space overlay.

Personally I rarely if ever use that mode but to support it you could simply look up to the canvas your on and if its in that mode use a different function for handling the input than if its not ... thats just off the top of my head I'm sure there is a more elegant way but as I basicly never use overlay I dont know :)

**Comment by theboyrossy** [comment]
Any chance of the complete edited code to work in World space? I'm having trouble getting your code snippet above to integrate with the existing code.

**Comment by ZinByakuya** [comment]
theboyrossy: I made the above code snippet work.

basically replace the old GetLocalMouse Method with the one provided by LodenDarkStar. the out mousePosition is a Vector2 so add a Vector2 mousePosition variable to the method.

then, every time it gets called (for example in the idle action) pass not a gameObject but a RectTransform.
(for example: hueGO.GetComponent<RectTransform>() [...])

if you only need it to work in the Screen Space - Camera Canvas Setting then this will work quickly, although this is not an overall solution.
